.buildkite/dagster-buildkite/dagster_buildkite/package_build_spec.py

```diff
@@ -0,0 +1,247 @@
+}
+
+PytestExtraCommandsFunction = Callable[[AvailablePythonVersion, Optional[str]], List[str]]
+PytestDependenciesFunction = Callable[[AvailablePythonVersion, Optional[str]], List[str]]
+
+
+class PackageBuildSpec(
+    NamedTuple(
+        "_PackageBuildSpec",
+        [
+            ("directory", str),
+            ("name", str),
+            ("package_type", str),
+            ("unsupported_python_versions", List[AvailablePythonVersion]),
+            ("pytest_extra_cmds", Optional[Union[List[str], PytestExtraCommandsFunction]]),
+            ("pytest_dependencies", Optional[Union[List[str], PytestDependenciesFunction]]),
+            ("pytest_tox_factors", Optional[List[str]]),
+            ("env_vars", List[str]),
+            ("tox_file", Optional[str]),
+            ("retries", Optional[int]),
+            ("upload_coverage", bool),
+            ("timeout_in_minutes", Optional[int]),
+            ("queue", Optional[BuildkiteQueue]),
+            ("run_pytest", bool),
+            ("run_mypy", bool),
+            ("run_pylint", bool),
+        ],
+    )
+):
+    """Main spec for testing Dagster Python packages using tox.
+
+    Args:
+        directory (str): Python directory to test, relative to the repository root. Should contain a
+            tox.ini file.
+        name (str, optional): Used in the buildkite label and coverage filename. Defaults to None
+            (uses the package name as the label).
+        package_type (str, optional): Used to determine the emoji attached to the buildkite label.
+            Possible values are "core", "example", "extension", and "infrastructure". By default it
+            is inferred from the location of the passed directory.
+        unsupported_python_versions (List[AvailablePythonVersion], optional): Python versions that
+            are not supported by this package. The versions for which pytest will be run are
+            the versions determined for the commit minus this list. If this result is empty, then
+            the lowest supported version will be tested. Defaults to None (all versions are supported).
+        pytest_extra_cmds (Callable[str, List[str]], optional): Optional specification of
+            commands to run before the main pytest invocation through tox. Can be either a list of
+            commands or a function. Function form takes two arguments, the python version being
+            tested and the tox factor (if any), and returns a list of shell commands to execute.
+            Defaults to None (no additional commands).
+        pytest_dependencies (Callable[str, List[str]], optional): Optional specification of
+            Buildkite dependencies (e.g. on test image build step) for pytest steps. Can be either a
+            list of commands or a function. Function form takes two arguments, the python version
+            being tested and the tox factor (if any), and returns a list of Buildkite step names.
+            Defaults to None (no additional commands).
```

Comment:
nit: "Optional specification of Buildkite dependencies" but being called `pytest_dependencies` is a bit confusing imo. `bk_step_dependencies` / `step_dependencies` maybe